File: src/main/java/agh/ics/oop/Vector2d.java
package agh.ics.oop;

public record Vector2d(int x, int y) {

    @Override
    public String toString() {
        return "(%d, %d)".formatted(this.x, this.y);
    }

    public boolean precedes(Vector2d other) {
        return other.x <= this.x && other.y <= this.y;
    }

    public boolean follows(Vector2d other) {
        return other.x >= this.x && other.y >= this.y;
    }

    public Vector2d upperRight(Vector2d other) {
        return new Vector2d(Math.max(this.x, other.x), Math.max(this.y, other.y));
    }

    public Vector2d lowerLeft(Vector2d other) {
        return new Vector2d(Math.min(this.x, other.x), Math.min(this.y, other.y));
    }

    public Vector2d add(Vector2d other) {
        return new Vector2d(this.x + other.x, this.y + other.y);
    }

    public Vector2d subtract(Vector2d other) {
        return new Vector2d(this.x - other.x, this.y - other.y);
    }

    public Vector2d opposite() {
        return new Vector2d(-1 * this.x, -1 * this.y);
    }

    public static void main(String[] args) {
        Vector2d position1 = new Vector2d(1, 2);
        System.out.println(position1);
        Vector2d position2 = new Vector2d(-2, 1);
        System.out.println(position2);
        System.out.println(position1.add(position2));
        System.out.println(position1.opposite());

        MapDirection west = MapDirection.WEST;
        System.out.println(west.next());
        System.out.println(west.previous());
        System.out.println(west.toUnitVector());
    }
}

File: KotlinZajecia/src/main/kotlin/Vector2D.kt
import kotlin.math.max
import kotlin.math.min

data class Vector2D(
    val x: Int,
    val y: Int
) {
    fun upperRight(other: Vector2D) = Vector2D(max(x, other.x), max(y, other.y))

    fun lowerLeft(other: Vector2D) = Vector2D(min(x, other.x), min(y, other.y))

    operator fun compareTo(other: Vector2D) = when {
        y != other.y -> (y - other.y)
        else -> (x - other.x)
    }

    operator fun unaryMinus() = Vector2D(-x, -y)

    operator fun plus(other: Vector2D) = Vector2D(x + other.x, y + other.y)

    operator fun minus(other: Vector2D) = Vector2D(x - other.x, y - other.y)
    operator fun rangeTo(other: Vector2D): List<Vector2D> {
        return (0 until other.x + 1)
            .flatMap { x ->
                (0 until other.y + 1)
                    .map { y -> Vector2D(x, y) }
            }
    }

}

fun MapDirection.toUnitVector(): Vector2D {
    return when (this) {
        MapDirection.NORTH -> Vector2D(0, 1)
        MapDirection.EAST -> Vector2D(1, 0)
        MapDirection.SOUTH -> Vector2D(0, -1)
        MapDirection.WEST -> Vector2D(-1, 0)
    }
}
